cocotb/drivers/amba.py

```diff
@@ -60,67 +85,164 @@ def __init__(self, entity, name, clock, **kwargs):
-        for cycle in range(delay):
-            yield RisingEdge(self.clock)
+            if hasattr(self.bus, "AWBURST"):
+                self.bus.AWBURST <= burst.value
 
-        self.bus.WDATA <= data
-        self.bus.WVALID <= 1
-        self.bus.WSTRB <= byte_enable
+            if hasattr(self.bus, "AWLEN"):
+                self.bus.AWLEN <= length - 1
 
-        while True:
-            yield ReadOnly()
-            if self.bus.WREADY.value:
-                break
-            yield RisingEdge(self.clock)
-        yield RisingEdge(self.clock)
-        self.bus.WVALID <= 0
-        self.write_data_busy.release()
+            if hasattr(self.bus, "AWSIZE"):
+                self.bus.AWSIZE <= size.bit_length() - 1
+
+            # Wait until acknowledged
+            while True:
+                await ReadOnly()
+                if self.bus.AWREADY.value:
+                    break
+                await RisingEdge(self.clock)
+            await RisingEdge(self.clock)
+            self.bus.AWVALID <= 0
+
+    @cocotb.coroutine
+    async def _send_write_data(self, data, delay, byte_enable, sync):
+        """Send the write data, with optional delay (in clocks)."""
+        async with self.write_data_busy:
+            if sync:
+                await RisingEdge(self.clock)
+
+            try:
+                byte_enable_iterator = iter(byte_enable)
+            except TypeError:
+                byte_enable_iterator = iter((byte_enable,))
+
+            for i, word in enumerate(data):
+                await ClockCycles(self.clock, delay)
+
+                self.bus.WDATA <= word
+                self.bus.WVALID <= 1
+
+                try:
+                    current_byte_enable = next(byte_enable_iterator)
+                    self.bus.WSTRB <= (2**self.bus.WSTRB.value.n_bits - 1
+                                       if current_byte_enable is None
+                                       else current_byte_enable)
+
+                except StopIteration:
+                    # Do not update WSTRB if we have reached the end of the
+                    # iterator
+                    pass
+
+                if hasattr(self.bus, "WLAST"):
+                    if i == len(data) - 1:
+                        self.bus.WLAST <= 1
+                    else:
+                        self.bus.WLAST <= 0
+
+                while True:
+                    await RisingEdge(self.clock)
+                    if self.bus.WREADY.value:
+                        break
+
+                if i == len(data) - 1:
+                    self.bus.WVALID <= 0
 
     @cocotb.coroutine
-    def write(
-        self, address: int, value: int, byte_enable: int = 0xf,
-        address_latency: int = 0, data_latency: int = 0, sync: bool = True
-    ) -> BinaryValue:
+    async def write(self, address, value, size=None, burst=AXIBurst.INCR,
+                    byte_enable=None, address_latency=0, data_latency=0,
+                    sync=True):
```

Comment:
Changing the argument order is a breaking change. Can you add the new arguments at the end?

Reply:
The arguments of AXI4LiteMaster.write() did not change, the extra arguments are available only on the new AXI4Master.write() method.

Reply:
Ah, apologies, I missed that.